utcp.c

```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
 		case FIN_WAIT_2:
 			// We shouldn't need to retransmit anything in this state.
 #ifdef UTCP_DEBUG
+			debug("Error: retransmit unexpected connection state %p %s\n", c, strstate[c->state]);
 			abort();
 #endif
-			timerclear(&c->rtrx_timeout);
-			break;
+			stop_retransmit_timer(c);
+			goto cleanup;
 	}
 
+	start_retransmit_timer(c);
+	utcp->rto *= 2;
+	if(utcp->rto > MAX_RTO)
+		utcp->rto = MAX_RTO;
+	c->rtt_start.tv_sec = 0; // invalidate RTT timer
+
+cleanup:
 	free(pkt);
 }
 
+/* Update receive buffer and SACK entries after consuming data.
+ *
+ * Situation:
+ *
+ * |.....0000..1111111111.....22222......3333|
+ * |---------------^
+ *
+ * 0..3 represent the SACK entries. The ^ indicates up to which point we want
+ * to remove data from the receive buffer. The idea is to substract "len"
+ * from the offset of all the SACK entries, and then remove/cut down entries
+ * that are shifted to before the start of the receive buffer.
+ *
+ * There are three cases:
+ * - the SACK entry is ahead of ^, in that case just change the offset.
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
+ */
+static void sack_consume(struct utcp_connection *c, size_t len) {
+	debug("sack_consume %lu\n", (unsigned long)len);
+
+	buffer_get(&c->rcvbuf, NULL, len);
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
+		if(len < c->sacks[i].offset) {
+			c->sacks[i].offset -= len;
+			i++;
+		} else if(len < c->sacks[i].offset + c->sacks[i].len) {
+			c->sacks[i].len -= len - c->sacks[i].offset;
+			c->sacks[i].offset = 0;
+			i++;
+		} else {
+			if(i < NSACKS - 1) {
+				// move remaining sacks one ahead
+				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
+				// clean last sack len
+				c->sacks[NSACKS - 1].len = 0;
```

Comment:
On Thu, Oct 29, 2015 at 08:02:30AM -0700, vollkorn1982 wrote:

> > +	for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
> > +		if(len < c->sacks[i].offset) {
> > +			c->sacks[i].offset -= len;
> > +			i++;
> > +		} else if(len < c->sacks[i].offset + c->sacks[i].len) {
> > +			c->sacks[i].len -= len - c->sacks[i].offset;
> > +			c->sacks[i].offset = 0;
> > +			i++;
> > +		} else {
> > +			if(i < NSACKS - 1) {
> > +				// move remaining sacks one ahead
> > +				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
> > +				// clean last sack len
> > +				c->sacks[NSACKS - 1].len = 0;
> 
> Hm. Let's say `NSACKS = 10` and `i = 5`. Now We end up in this branch, move all the sacks to the left, `i` is still 5 and `c->sacks[NSACKS - 1]` resolves to `c->sacks[9].len = 0`. Then we loop again and the for condition is checked: `c->sacks[5].len` did change to the len of the sack which was previously one to the right. We go into this branch again and again move the same amount of sacks one to the left, again copying the last one, making sure its len is set to 0 and thus propagating the len = 0 to the left until it hits position 5. All the sacks [5, NSACKS[ are effectively removed. Isn't there an easier, more straight forward way to just toss out the remaining sacks?

I set NSACKS to 4 after testing how many sack regions were created when
having 20% packet loss. At this level, because of the triplicate ACK
mechanism, the sender will almost certainly start retransmission before
the receiver would have to use more SACK regions. So I wouldn't
increase NSACKS at all. With only 4 sack entries, I don't think you'd
gain anything by adding logic to detect if you can optimize that last
branch.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>